update unmanaged resources compilation script

guimou opened the pull request:
Replace environment variables to store temp values by script variables and temporary files. Needed in case the size of the env var exceeds limits, which can happen with really long descriptions.

<!--- Provide a general summary of your changes in the Title above -->

## Description
Modification to the script compiling the unmanaged sources to update the configmap. I replaced environment variables to store temp values by script variables and temporary files. This is needed in case the size of the env var exceeds limits, which can happen with really long descriptions.

## How Has This Been Tested?
Tested in a real deployment.

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
- All the commits have been [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits)  (To pass the `DCO` check)

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] The commits have meaningful messages; the author will squash them after approval or in case of manual merges will ask to merge with squash.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has manually tested the changes and verified that the changes work.
- [X] Code changes follow the [kubeflow contribution guidelines](https://www.kubeflow.org/docs/about/contributing/).
- [X] **For first time contributors**: Please reach out to the [Reviewers](../OWNERS) to ensure all tests are being run, ensuring the label `ok-to-test` has been added to the PR.

If you have UI changes

<!--- You can ignore these if you are doing Go Model Registry REST server, MR Python client, manifest, controller, internal logic, etc changes; aka non-UI / visual changes -->
- [ ] The developer has added tests or explained why testing cannot be added.
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Verify that UI/UX changes conform the UX guidelines for Kubeflow.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Chores**
	- Improved the process for updating configuration by switching to a more reliable file-based approach when handling JSON content.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tarilabs commented on clients/python/src/model_registry/_client.py:
```diff
@@ -0,0 +1,148 @@
+        self._api.upsert_model_artifact(ma, mv.id)
+        return ma
+
+    def register_model(
```
should we rename this ~"register model artifact" or something else to better indicate this is performing the whole chain of inserting (or reusing) RegisteredModel, ModelVersion and ModelArtifact? 🤔

lampajr replied:
In this regard IMO `register_model` is good enough, I found the same pattern on [mlflow api](https://mlflow.org/docs/latest/model-registry.html#api-workflow):
- `mlflow.register_model` high level api to register the full model (model, version and artifact)
- `client.create_registered_model` raw api to create empty model - the same as our `upsert_registered_model`
- `client.create_model_version` raw api to create a new model version - the same as our `upsert_model_version`

wdyt?

tarilabs replied:
yes, thank you for making me notice that, lgtm 👍

tarilabs reviewed (approved):
minor comment which I believe is better to address before merging, but I don't want to hold merging solely based on that.

Thank you so much this will be extremely helpful

lampajr commented on clients/python/src/model_registry/__init__.py:
```diff
@@ -2,11 +2,8 @@
 
-from .client import ModelRegistry
-from .types import ListOptions, OrderByField
+from ._client import ModelRegistry
```
So if I correctly understood, with this we are going to "export" just the high level API and not the low level one (i.e., `ModelRegistryAPIClient`) - this way users cannot directly use the underlying API, right?

If I am right, is this what we want? I am thinking how user could update for instance a model artifact metadata if we are not exposing the underlying.

If I am not right, forget what I said :)

isinyaaa replied:
Yes, by default we only export the higher-level API, but the user can also do `from model_registry.core import ModelRegistryAPIClient as MRAPI` and use the lower-level bindings the same way they would in go.

lampajr replied:
Well actually if the user can import the lower level api that is fine for me, make sense to export by default just the higer level one.

But we should guarantee a way to make use of the lower level if the user want to use it, at least for now I think

tarilabs replied:
> But we should guarantee a way to make use of the lower level if the user want to use it, at least for now I think

+1

Thanks for the clarifications Isabella

guimou committed:
update unmanaged resources compilation script

Replace environment variables to store temp values by script variables and temporary files.
Needed in case the size of the env var exceeds limits, which can happen with really long descriptions.

Signed-off-by: Guillaume Moutier <guillaume.moutier@gmail.com>

tarilabs commented:
adding @mturley for awareness, thanks @guimou

adysenrothman reviewed (commented):
/lgtm

Tested this locally by adding bulk to the README md for the models. I ran the update script to find that the updated configmap is stored in json format as expected. If the update fails, the rm of the temp json file can be commented out to save the new configmap if needed. 

I'm curious what the fix is if the mem limit is exceeded? I attempted to manually update the configmap YAML via the OpenShift console but was unable to (getting the same limit error which I believe would be expected). 

Could you provide more clarity on this? Is this a fixable issue or will these configmaps need to be broken into smaller pieces to resolve this?

tarilabs reviewed (approved):
/approve

guimou commented:
@adysenrothman You're right, there is a limit of size for a ConfigMap (same for a Secret) and this may be a problem in some cases (default recommended is 1.5MB, with generally a hard stop at 3MB).
A few possible solutions:
- Store the JSON file in PVC instead. This would require other mechanisms to do the upload and read it. Not my favourite solution.
- A set of ConfigMaps, one per model, with a specific annotation so that the Catalog knows it has to handle them.
- Yet another CRD specifically for that, like "ModelCatalogItem", with one instance per model. This is the cleanest and the most Kubernetes way in my opinion. But it's another CRD to create/manage.

adysenrothman commented:
> @adysenrothman You're right, there is a limit of size for a ConfigMap (same for a Secret) and this may be a problem in some cases (default recommended is 1.5MB, with generally a hard stop at 3MB). A few possible solutions:
> 
> * Store the JSON file in PVC instead. This would require other mechanisms to do the upload and read it. Not my favourite solution.
> * A set of ConfigMaps, one per model, with a specific annotation so that the Catalog knows it has to handle them.
> * Yet another CRD specifically for that, like "ModelCatalogItem", with one instance per model. This is the cleanest and the most Kubernetes way in my opinion. But it's another CRD to create/manage.

Thanks for the clarification! I also appreciate these suggestions. Will keep this in mind moving forward as we reconfigure the MC to be more flexible.
